src/main/java/build/buildfarm/cas/cfc/CASAccessMetricsRecorder.java

```diff
@@ -0,0 +1,242 @@
+package build.buildfarm.cas.cfc;
+
+import static java.lang.String.format;
+
+import build.bazel.remote.execution.v2.Digest;
+import build.buildfarm.backplane.Backplane;
+import build.buildfarm.common.DigestUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Level;
+import lombok.SneakyThrows;
+import lombok.extern.java.Log;
+
+/**
+ * This class records read and write metrics, updating the information to central storage at regular
+ * intervals. The collected data can be utilized to determine the read count of any CAS entry, query
+ * the top "n%" of frequently accessed CAS entries, and rank CAS entries based on their read
+ * frequency, among other uses.
+ */
+@Log
+public final class CASAccessMetricsRecorder {
+  private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
```

Comment:
This scheduled executor service should be injected.

Reply:
Done